Enable QtWebEngine modules for Qt6 package on Artifactory

cedrik-fuoco-adsk opened the pull request:
This PR is about enabling QtWebEngine modules for the Qt6 package in Artifactory. This PR contains 5 different commits that could be split into multiple PRs if needed, or each commit could be looked at independently for review. 

All 5 commits are needed to support QtWebEngine.
**Let me know if I should split the PR.** 

------
The [first commit](https://github.com/AcademySoftwareFoundation/aswf-docker/commit/ded19f38f422ab5c9841bb83b3eca47220aa997a) adds all the new Conan recipes without any changes. 
Here's the list (the order is based on which one need to be built first):
- m4
- autoconf
- automake
- flex
- bison
- expat
- meson
- pkgconf
- dbus
- gnu-config
- libtool
- xorg-macros
- libpciaccess
- linux-headers-generic
- nspr
- sqlite3
- zlib
- nss
- opus
- xorg-proto
- libxshmfence
- libdrm
- gperf
- nodejs

I know that there are a few dependencies that are already installed with `install_yumpackages.sh`, but s**hould we use the version that comes from yum/dnf packages or gradually move what we can to Conan?**

------

The [second commit](https://github.com/AcademySoftwareFoundation/aswf-docker/commit/5c5b4116300a0bd6e0ef361aa421327047dbe303) is adding the copyrights notice and changing the versions to `ASWF_<PACKAGE>_VERSION` environment variable.

------

The [third commit](https://github.com/AcademySoftwareFoundation/aswf-docker/commit/0bfca1e7cb5a322bdfbdcd670d05c738764e0814) is changing `builder.py` and the `ci-image-dockerfile` template to allow Conan packages that have hyphen in their name. It was causing issue with the environment variables because hyphen is not a allowed character for environment variable name.

------

The [fourth commit](https://github.com/AcademySoftwareFoundation/aswf-docker/commit/51cddacd108c9fe23532eec30bd58f65a6a05917) is moving the installation of html5lib to pip because when using yum/dnf, it install the html5lib for the system Python - which is not the right one. It was causing issue during the build for WebEngine because it was not able to find html5lib.

------

The [fifth commit](https://github.com/AcademySoftwareFoundation/aswf-docker/commit/8353d61fceaded5dc9381595a9fdd3aa59d8e673) is doing the following:
- Adding all the new environment variables to `packages/common/Dockerfile`
- Adding all the new packages to versions.yaml
- Modifying Qt recipe to use the newly added package
- Edited VFX2024 profile to enable all modules that are needed for QtWebEngine. These new options only affects Qt recipe.

**I've added all the new packages to 2024.1, let me know if it should be 2024.2.**

jfpanisset commented:
Hi Cedrik, thank you so much for this, I'll take a closer look over the weekend.

One question I had from a cursory look: a lot of the Conan packages being added as dependencies for Qt are already in the base Rocky 8 image (flex / bison / automake / ...). Until now the approach has been that only packages for which Rocky 8 does not have an RPM get added as Conan packages: otherwise I just comment out the explicit dependency and let the build find the system installed version. Of course the exception would be if the package being built absolutely requires a newer version than what's included wth RHEL 8 / Rocky 8.

Basically I want to avoid "building an OS distribution", and keeping the build environment as close as possible to what someone might have on a "stock" RHEL 8 / Rocky 8 / Alma 8 machine.

I'll have more specific comments over the weekend, thanks again!

cedrik-fuoco-adsk commented:
I've removed all the Conan packages except `expat`, and moved them to `install_yumpackages.sh`.

I am currently compiling the Qt6 package. The configuration look good, but I am waiting for the compilation to finish to confirm that QtWebEngine and QtMultimedia are there.

For the future, I kept the changes to handle dependencies with hyphens in it. Let me know if you prefer to not include that.

-----

The next step will be to work on `ci-openrv`. It is mostly working already (locally), but I have to check that all the packages needed for OpenRV are there. Unfortunately, OpenRV won't compile since Qt6 is not supported yet.

cedrik-fuoco-adsk commented on scripts/common/install_yumpackages.sh:
I just realised that I added dependencies to build other direct dependencies for QtWebEngine and QtMultimedia.
I will clean those up on Monday. Here are the direct dependencies for QtWebEngine and QtMultimedia:

**QtWebEngine:**
- expat
- opus
- xorg-proto
- libxshmfence
- nss
- libdrm
- dbus
- nodejs
- gpref
- bison
- flex

**QtMultimedia:**
- libalsa
- openal

jfpanisset commented on python/aswfdocker/data/versions.yaml:
```diff
@@ -687,6 +688,8 @@ ci-packages:
     - "2023.2"
     - "2024.1"
+  expat:
+    - "2024.1"
```
I have no idea how the `ci-packages` section is sorted,  if at all. As the newest package, at the end is fine.

jfpanisset commented on python/aswfdocker/data/versions.yaml:
```diff
@@ -828,6 +831,7 @@ groups:
         - gtest
         - blosc
       base2:
+        - expat
```
Not a blocker, but I would consider `expat` a more generic dependency than just for qt, for instance OCIO looks for it in its `Config.cmake.in`, so I'd move expat to the `base1` build stage.

AFAIK the reason why `qt` and `pyside` are in separate `base2` and `base3` sections is due to the build time which surpases what can be done on a base GHA runner. `CONTRIBUTING.md` documents how to upload a release created from a local build, but now that we have access to larger runners, it is preferable to build everything using GHA.

cedrik-fuoco-adsk replied:
I agree, I can move it to base1.

jfpanisset commented on scripts/common/install_pippackages.sh:
```diff
@@ -0,0 +1,15 @@
+    # There is a package called python3-html5lib through yum/dnf but it install it
+    # for the system Python - which is not the same one that it used by downstream projects.
+    python -m pip install html5lib
+fi
```
As discussed above, this adds a non-version source of dependencies. Also `html5lib` is a bit hidden in this install script. Would it make sense to have the script be more generic and install from a `requirements.txt` file?

cedrik-fuoco-adsk replied:
I agree that it is a bit hidden. We can also pin the version within the command itself (eg. `python -m pip install html5lib==1.1`), but it would still be hidden.

Were you thinking of adding a requirements.txt at the root of the repository? @jfpanisset

jfpanisset reviewed (approved):
This is great stuff, thank you so much for contributing this. None of the comments I've made are blockers, so I'd rather merge as is so you can keep progressing and I'll address a few in an upcoming branch to finalize the 2024.1 releases.

jfpanisset commented on packages/conan/recipes/expat/conanfile.py:
```diff
@@ -0,0 +1,114 @@
+        cmake = CMake(self)
+        cmake.install()
+        rmdir(self, os.path.join(self.package_folder, "lib", "pkgconfig"))
+        rmdir(self, os.path.join(self.package_folder, "lib", "cmake"))
```
Also we typically want to keep the `cmake` directory so packages can call CMake on Conan-installed packages outside the context of Conan, so typically line 88 would be commented out.

cedrik-fuoco-adsk commented on packages/common/Dockerfile:
```diff
@@ -133,6 +139,9 @@ ENV ASWF_PARTIO_VERSION=${ASWF_PARTIO_VERSION}
 
+ARG ASWF_EXPAT_VERSION
+ENV ASWF_EXPAT_VERSION=${ASWF_EXPAT_VERSION}
+
```
I kept expat as a Conan package because the Qt recipe was asking for 2.5.0. I did not test if an earlier version of expat would work though.

cedrik-fuoco-adsk commented:
> This is great stuff, thank you so much for contributing this. None of the comments I've made are blockers, so I'd rather merge as is so you can keep progressing and I'll address a few in an upcoming branch to finalize the 2024.1 releases.

I haven't look at the expat recipe in details. I'll go over the changes and I can do another PR for that.

cedrik-fuoco-adsk commented on packages/conan/settings/profiles/vfx2024:
```diff
@@ -3,5 +3,13 @@ include(ci_common4)
+qt:qtwebchannel=True
+qt:with_dbus=True
+qt:qtwebsockets=True
+qt:qtmultimedia=True
```
I was following the "Conan" way of enabling options, but since we are already modifying the Qt recipe, I could set the defaults to `true` within the recipe directly (they are all set to `false` by default).

cedrik-fuoco-adsk commented on python/aswfdocker/builder.py:
```diff
@@ -105,7 +105,8 @@ def make_bake_dict(
                             "ASWF_"
-                            + image.replace("ci-package-", "").upper()
+                            # Handle dependencies with hyhpen in their name.
+                            + image.replace("ci-package-", "").upper().replace("-", "_")
```
Yes, it was needed when I had these dependencies. As you said, it is not needed right now, but I think it is a good thing to have too.
